base/genericmemory.jl

```diff
@@ -25,8 +37,9 @@ Memory
 """
     AtomicMemory{T} == GenericMemory{:atomic, T, Core.CPU}
 
-One-dimensional dense array with elements of type `T`, where each element is
+One-dimensional, fixed-size, dense array with elements of type `T`, where each element is
 independently atomic when accessed, and cannot be set non-atomically.
+For details, see [Atomic Operations](@ref man-atomic-operations)
```

Comment:
what does it mean precisely in terms of julia code? `man-atomic-operations` deals with `Threads.Atomic` as well as `@atomic` macro. I presume these sentences relate to the latter. Does that mean that `at_mem[3] = 5` is performed atomically (without adding `@atomic` in front? Same question for `x = at_mem[3]`. 

I've been waiting for using this already quite some time but it's really hard without any docs ;) e.g. can a non-isbits `x` be in a non-consistent state when read from `at_mem::AtomicMemory` (e.g. some other thread was modifying it at the same time)?
I presume that the whole point of `AtomicMemory` is that one does not need to use `@atomic`, but I'd be nice to have it stated in clear text here.

also a tangent question, but the next building block would be a CAS operation on an `AtomicMemory`, but neither this docstring nor `man-atomic-operations` mention it (except for `Threads.atomic_cas!` which deals only with `Threads.Atomic`).

Reply:
does that mean that `@atomic` macro will be obligatory for reading `AtomicMemory`? or that it would default to `:monotonic`? how about writing into `AtomicMemory`? and of course cas operations: will there be `@atomiccas mem[3] 4=>2` ala Atomix.jl?

Reply:
reads default to monotonic. writes require an explicit order. not sure about CAS.

Reply:
@oscardssmith it would be great if this information were included in the docs!

Reply:
agreed. it's one that's easier to document once we actually give it the correct interface. Currently there's no public way to use an AtomicMemory which makes it slightly tricky to document.

Reply:
which makes it a chicken and egg problem ;) I'd like to experiment with it, so I'd need a minimal version of working docs, which are not there due to the experimental nature of it, which is very hard to finalize without people experimenting with it ;)

I'll add a suggestion for the docs to include what you said above.

Reply:
The order of parameters to the Builtins are documented: https://docs.julialang.org/en/v1.12-dev/devdocs/builtins/#Core.memoryrefget

And some of the meaning of the functions are described here, in the plan: https://hackmd.io/5Cv87gEhSVWI84uriGVxyA#New-Builtin-Intrinsic-Functionality
and via cross-reference to the the generic function's docs: https://docs.julialang.org/en/v1.12-dev/base/base/#Base.swapproperty!

Reply:
@vtjnash thanks for the info. The Manifesto was modified last time one year ago and doesn't even mention `AtomicMemory` so I was hesitant to rely on the document.

What I understand is that the plan is to implement
* `@atomic ord x[3]` → `Core.memoryrefget(memoryref(x, 3), ord, ...)`
* `@atomic ord x[3] = 4` → `Core.memoryrefset!(memoryref(x, 3), 4, ord, ...)`
* `@atomicswap ord x[3]` → `Core.memoryrefswap!(memoryref(x, 3), 4, ord, ...)`
* `@atomicreplace ord x[3] 3=>4` → `Core.memoryrefreplace!(memoryref(x, 3), 3, 4, ord, ...)`

to extend the `@atomic[...]` macros? (So far these work only on atomic fields). I'd happily give it a try myself if that's the goal ;)

Reply:
also docstrings of `Core.memoryref[...]` functions mention often `isatomic` parameter which I infer to be the type parameter of `AtomicMemory` but `isatomic` sounds more like function (on `Memory`?) and is not mentioned at all in the docs elsewhere.

Reply:
I believe that's a remnant of an earlier version of the pr where we had an isatomic bool for the type parameter rather than a symbol kind. time to fix the docs more :)

Reply:
For CAS, we have `Core.memoryrefreplace!` and `Core.memoryrefmodify!` which are the implimentations of the (to be created `modifyindex!` and `replaceindex!` that function analogously to `modifyfield!` and `replacefield!` respectively.

Reply:
Also when we do this, we will want to give CUDA.jl a companion PR that updates https://github.com/JuliaGPU/CUDA.jl/blob/b07cf990bce048e106c7d767ae6003d574706407/src/device/intrinsics/atomics.jl to use the new names.